### PR TITLE
ENH: Initial commit for RTIonBeam GUI elements, RTPlan node rework and ...

### PR DESCRIPTION
**Pull request by MichaelColonel**
...other minor enhancements

### CODE REVIEW

File: Beams/Widgets/qMRMLBeamParametersTabWidget.cxx [hunk 1]
```diff
@@ -149,12 +153,14 @@ void qMRMLBeamParametersTabWidget::setBeamNode(vtkMRMLNode* node)
...
 
   vtkMRMLRTBeamNode* beamNode = vtkMRMLRTBeamNode::SafeDownCast(node);
+  vtkMRMLRTIonBeamNode* ionBeamNode = vtkMRMLRTIonBeamNode::SafeDownCast(beamNode);
 
   // Connect display modified event to population of the table
   qvtkReconnect( d->BeamNode, beamNode, vtkCommand::ModifiedEvent,
                  this, SLOT( updateWidgetFromMRML() ) );
 
   d->BeamNode = beamNode;
+  d->IonBeamNode = ionBeamNode;
```
Comment (cpinter) [inline Beams/Widgets/qMRMLBeamParametersTabWidget.cxx]: I'd strongly prefer storing only one pointer (keep BeamNode and remove IonBeamNode) and doing SafeDownCast at the places of usage to decide what kind of beam it is.

File: ExternalBeamPlanning/Logic/vtkSlicerExternalBeamPlanningModuleLogic.cxx [hunk 2]
```diff
@@ -200,7 +200,7 @@ void vtkSlicerExternalBeamPlanningModuleLogic::OnMRMLSceneEndImport()
   // Observe plan events of all plan nodes
   std::vector<vtkMRMLNode*> nodes;
   this->GetMRMLScene()->GetNodesByClass("vtkMRMLRTPlanNode", nodes);
-  for (std::vector<vtkMRMLNode*>::iterator nodeIt=nodes.begin(); nodeIt!=nodes.end(); ++nodeIt)
+  for ( auto nodeIt=nodes.begin(); nodeIt!=nodes.end(); ++nodeIt)
```
Comment (cpinter) [inline ExternalBeamPlanning/Logic/vtkSlicerExternalBeamPlanningModuleLogic.cxx:203]: Please don't add space after parentheses. At line 214 as well.

**Comment by cpinter** [comment]
I have not tested the branch yet, but it looks good, except a few minor comments I left. @Sunderlandkyl any thoughts?

**Comment by Sunderlandkyl** [comment]
LGTM!

File: Beams/Widgets/qMRMLBeamParametersTabWidget.cxx (hunk 1 above)
Comment (MichaelColonel) [reply Beams/Widgets/qMRMLBeamParametersTabWidget.cxx]: Only one pointer left.

File: ExternalBeamPlanning/Logic/vtkSlicerExternalBeamPlanningModuleLogic.cxx (hunk 2 above)
Comment (MichaelColonel) [reply ExternalBeamPlanning/Logic/vtkSlicerExternalBeamPlanningModuleLogic.cxx:203]: Code style fixed.

File: Beams/Widgets/qMRMLBeamParametersTabWidget.cxx [hunk 3]
```diff
@@ -261,6 +266,62 @@ void qMRMLBeamParametersTabWidget::updateWidgetFromMRML()
       }
     } // For each row
   } // For each tab
+
+  // Update RTBeamNode specific widgets according to the type of rt beam node
+  vtkMRMLRTIonBeamNode* IonBeamNode = vtkMRMLRTIonBeamNode::SafeDownCast(d->BeamNode);
```
Comment (Sunderlandkyl) [inline Beams/Widgets/qMRMLBeamParametersTabWidget.cxx]: First letter in IonBeamNode should be lowercase.

File: Beams/MRML/vtkMRMLRTPlanNode.cxx [hunk 4]
```diff
@@ -107,35 +104,15 @@ void vtkMRMLRTPlanNode::ReadXMLAttributes(const char** atts)
...
-    {
-      this->RxDose = vtkVariant(attValue).ToDouble();
-    }
-    else if (!strcmp(attName, "IsocenterSpecification"))
-    {
-      this->IsocenterSpecification = (IsocenterSpecificationType)(vtkVariant(attValue).ToInt());
-    }
-  }
+  vtkMRMLReadXMLBeginMacro(atts);
+  vtkMRMLReadXMLIntMacro( NextBeamNumber, NextBeamNumber);
```
Comment (Sunderlandkyl) [inline Beams/MRML/vtkMRMLRTPlanNode.cxx]: Ctrl+F still shows 47 different places with spaces following parenthesis.
Could you run a  find+replace to fix them?

Comment (MichaelColonel) [reply Beams/MRML/vtkMRMLRTPlanNode.cxx]: I've corrected a format.

**Comment by cpinter** [comment]
Thanks @MichaelColonel! Apparently all the issues are solved, so once you squash the commits, I'll merge.
Btw in the future if you amend your initial commit based on the feedback, then there is no need for this extra step.

**Comment by MichaelColonel** [comment]
Squashed to one commit.

**Comment by cpinter** [comment]
Awesome, thanks! Merged.

**Pull request merged**